### PR TITLE
fix(ifstep audit trail): fix of ifstep audit trail and context when we use nestedsteps

### LINKED ISSUE #51: IfStep with nested steps auditTrail bug
**Describe the bug**
When we use the If Step and the step then or else is a list of steps, it works, but in the auditTrail don't have any return in the Ok, just errors.

**To Reproduce**

1. Create an ifstep like this:
```
"Description of conditional rule": ifElse({
    "Conditional logic": step(async (ctx) => {
        if (ctx.right) return Ok(true)
        return Ok(false)
    }),

    "Then: True condition": step(async (ctx) => {
        return Ok()
    }),

    "Else: False condition": step({
        "Description of step 1": step((ctx) => {
            const myValueTest = true
            return Ok(myValueTest)
        }),
        "Description of step 2": step((ctx) => {
            const myValueTest2 = { myOjbect: "some string" }
            return Ok(myValueTest2)
        })
    })
})
```

2. Execute the useCase, for this example I will put the requisite to fall on the else flow

3. Then we go see the auditTrail that we recept something like this:
```
description:'Description of conditional rule',
elapsedTime:440800n,
returnElse: { 
    isOk: true,
    ok :{}
},
returnIf: {
    Ok :{value: false}
},
type:'if else'
```

When we use the ifstep with nested steps, the buchu should came with steps and individual return, don't?

### PR DESCRIPTION
**Pull request by endersoncosta**
fix #51

## Proposed Changes

1. If we use nestedsteps without external context, the _stopExecution don't exist, therefore then if we have context, the internal context will be overwrite, else we can use the internal context just to don't break the application.
2. Today when we execute an IfElse step the audit trail is lost, because we use the return of steps to compose the ifstep audit trail, I changed this to insert the audit trail corretly.

### CODE REVIEW

File: test/ifElse/ifElseStep.js [hunk 1]
```diff
@@ -368,4 +409,306 @@ describe('If Else step', () => {
...
+                            }
+                        ]
+                    },
+                    elapsedTime: st.auditTrail.elapsedTime
+                })
+                assert.ok(st.auditTrail.elapsedTime > 0)
+            })
+        })
+
+        context('returning Ok from Else', () => {
```
Comment (dalssoft) [inline test/ifElse/ifElseStep.js:536]: it is returning `Err`, not a `Ok` from `Then`

File: test/ifElse/ifElseStep.js [hunk 2]
```diff
@@ -368,4 +409,306 @@ describe('If Else step', () => {
...
+                        ]
+                    },
+                    elapsedTime: st.auditTrail.elapsedTime
+                })
+                assert.ok(st.auditTrail.elapsedTime > 0)
+            })
+        })
+    })
+
+    describe('using nested steps with context', () => {
```
Comment (dalssoft) [inline test/ifElse/ifElseStep.js:601]: how `using nested steps with context` test is different from `using nested steps`?

**Review by dalssoft — CHANGES_REQUESTED** [review]

File: test/ifElse/ifElseStep.js (hunk 2 above)
Comment (endersoncosta) [reply test/ifElse/ifElseStep.js:601]: Nestedsteps use the _stopExecution which came from the context, if we overwrite the context with undefined, then the nestedstep will break.

Seeing this, I put an if and made two tests, with context and without context.

File: test/ifElse/ifElseStep.js (hunk 1 above)
Comment (endersoncosta) [reply test/ifElse/ifElseStep.js]: fixed

File: test/ifElse/ifElseStep.js (hunk 2 above)
Comment (dalssoft) [reply test/ifElse/ifElseStep.js:601]: Sorry, but I could not find the difference between these two test cases. It seems they are exercising the same code / scenario. Could you spot to me the difference?

File: test/ifElse/ifElseStep.js (hunk 1 above)
Comment (dalssoft) [reply test/ifElse/ifElseStep.js]: Shouldn't we have
`returning Ok from Then`
`returning Err from Then`
`returning Ok from Else`
`returning Err from Else` ?

**Review by dalssoft — CHANGES_REQUESTED** [review]

File: test/ifElse/ifElseStep.js (hunk 2 above)
Comment (endersoncosta) [reply test/ifElse/ifElseStep.js:601]: Right, in the ifElse step, we had `step.context = this.context` without if.

With this code in this manner, if we performed that code with usecase context, then will work correctly. But if we run the code without usecase context, the context of the step don't has the _stopExecution and will break the application...

I included an If statement to fix this , then this affects the two flows, for this reason I choose run the "same test" to both contexts.

File: test/ifElse/ifElseStep.js (hunk 1 above)
Comment (dalssoft) [reply test/ifElse/ifElseStep.js]: Sorry but I couldn't find the explanation why this was marked as resolved.

Comment (endersoncosta) [reply test/ifElse/ifElseStep.js:658]: oh sorry, I couldn't understand at that time, right now I fixed the correctly the test.

**Comment by jhomarolo** [comment]
@endersoncosta is the PR name using the correct semantics? https://github.com/herbsjs/buchu/blob/master/.github/CONTRIBUTING.md

**Comment by endersoncosta** [comment]
> @endersoncosta is the PR name using the correct semantics? https://github.com/herbsjs/buchu/blob/master/.github/CONTRIBUTING.md

I used the commitizen... And in the contributing.md I can't understand if has any pattern to this matter.

**Comment by dalssoft** [comment]
> And in the contributing.md I can't understand if has any pattern to this matter.

Agree

**Comment by endersoncosta** [comment]
@dalssoft what's left to finish this case?

**Comment by dalssoft** [comment]
@endersoncosta there is an open issue

>Given there is no difference or new input, the test should be just one. Or force the input to be different.

**Comment by endersoncosta** [comment]
Right, the sceneries was different, but I made some complement to add more sceneries with context...

**Comment by endersoncosta** [comment]
the test has been fixed

**Review by dalssoft — APPROVED** [review]

**Pull request merged**